docs/architecture/gas/estimator.md

```diff
@@ -20,8 +20,135 @@ clients.
+
+```bash
+cargo run -p runtime-params-estimator --features required -- \
+    --accounts-num 20000 --additional-accounts-num 20000 \
+    --iters 3 --warmup-iters 1 --metric time \
+    --costs=ActionReceiptCreation,ActionTransfer,ActionCreateAccount,ActionFunctionCallBase
+```
+
+You should get an output like this.
+
+```
+[elapsed 00:00:17 remaining 00:00:00] Writing into storage ████████████████████   20000/20000  
+ActionReceiptCreation         4_499_673_502_000 gas [  4.499674ms]    (computed in 7.22s) 
+ActionTransfer                  410_122_090_000 gas [   410.122µs]    (computed in 4.71s) 
+ActionCreateAccount             237_495_890_000 gas [   237.496µs]    (computed in 4.64s) 
+ActionFunctionCallBase          770_989_128_914 gas [   770.989µs]    (computed in 4.65s) 
+
+
+Finished in 40.11s, output saved to:
+
+    /home/you/near/nearcore/costs-2022-11-11T11:11:11Z-e40863c9b.txt
+```
+
+This shows how much gas a parameter should cost to satisfy the 1ms = 1Tgas rule.
+It also shows how much time that corresponds to and how long it took to compute
+each of the estimations.
+
+Note that the above does not produce very accurate results and it often has high
+variance as well. It runs an unoptimized binary, the database is small, and the
+metric used is wall-clock time.
+
+Once you have your estimation code all ready, it is better to run with large state and an optimized binary.
+
+```bash
+cargo run --release -p runtime-params-estimator --features required -- \
+    --accounts-num 20000 --additional-accounts-num 2000000 \
+    --iters 3 --warmup-iters 1 --metric time \
+    --costs=ActionReceiptCreation,ActionTransfer,ActionCreateAccount,ActionFunctionCallBase
+```
+
+You might also want to run a hardware-agnostic estimation using the following
+command. It uses `docker` and `qemu` under the hood, so it will be quite a bit
+slower. It also won't work if `docker` is not installed on your system.
+
+
+```bash
+cargo run --release -p runtime-params-estimator --features required -- \
+    --accounts-num 20000 --additional-accounts-num 2000000 \
+    --iters 3 --warmup-iters 1 --metric icount --docker --full \
+    --costs=ActionReceiptCreation,ActionTransfer,ActionCreateAccount,ActionFunctionCallBase
+```
+
+Note how the output looks a bit different now. The `i`, `r` and `w` values show
+instruction count, read IO bytes and write IO bytes. The IO byte count is known
+to be inaccurate.
+
+```
++ /host/nearcore/runtime/runtime-params-estimator/emu-cost/counter_plugin/qemu-x86_64 -plugin file=/host/nearcore/runtime/runtime-params-estimator/emu-cost/counter_plugin/libcounter.so -cpu Westmere-v1 /host/nearcore/target/release/runtime-params-estimator --home /.near --accounts-num 20000 --iters 3 --warmup-iters 1 --metric icount --costs=ActionReceiptCreation,ActionTransfer,ActionCreateAccount,ActionFunctionCallBase --skip-build-test-contract --additional-accounts-num 0 --in-memory-db
```

Comment:
Long term future work: should we consider what set of machines we want to support and run regular CI testing on them.

Reply:
Not sure what you mean.

Since the comment is on the qemu command, do you mean to run it with different CPU targets in qemu? I don't think qemu is even trying to reflect performance differences between CPUs, so I am skeptical on how useful that would be.

Or do you mean to set up multiple "real" machines with different hardware? That has been discussed as long-term goal before. Right now we have continuous estimation (CE) that runs it on "a" machine. We already struggle to make that vaguely stable across runs. We would probably need single tenant nodes, disable frequency boosting, and eliminate all other sources of variance to make progress there.
If and when we have managed to do that for a single machine, I would really like to extend it to cover different hardware. But we haven't assigned priority since I finished the initial CE work.

Reply:
Sorry, I now realise my original comment was not very clear.  I really meant OSes and not machines.  We should figure out which OSes protocol contributors are typically expected to use and make sure that the above commands work on them.  We could start with e.g. ubuntu and Mac.

Reply:
Ah, I see. This specific command runs inside docker, so I think it should be fine with any OS?
For everything else (not docker + qemu runs), I don't see how this is specific to the estimator.

But yeah I think the discussion to run proper CI on Mac has come up before and I think it makes sense to have that.

Reply:
> Ah, I see. This specific command runs inside docker, so I think it should be fine with any OS?

Yeah, I suppose it will boil down to simply is docker supported on the set of OSes we want to support.

Reply:
On a side note, the long-term goal in my head is to 50% to remove the qemu and docker setup entirely, as it is a substantial maintenance burden. But it's quite a long way still before we could do that, because we don't have a good alternative estimation method that is stable enough.

The other 50% of me that think we want to keep it as a second line of defense capturing regression. And perhaps for performance debugging where exact counts can sometimes be useful.